### PR TITLE
278153 voltas

### PR DESCRIPTION
**Pull request by pthvogt**
We move the layout of the voltas down in both the normal layout and in linear layout.

Fix for #278153: We have to move layout of the voltas after the tempo text layout, as the layout of the volta sets also the tempo. This is taken from the system, but set in the layout of the tempo text. If we layout the voltas before the tempo text, we set the wrong tempo at the end of the voltas.

Fix for #279391: We have to move the the layout after the layout of RehearsalMark, StaffText, FretDiagram, so that these elements go underneath the volta.

### CODE REVIEW

**Comment by pthvogt** [comment]
I tried my best :-)

OK this way?

**Comment by anatoly-os** [comment]
No :) You should have got 1 commit in this PR. I suggest using GitKraken to visualise what you have in the repo trees. You can easily revert changes and make all operations, but pushing changes there.

**Comment by pthvogt** [comment]
Next try :-)

(git is a beast.)

**Comment by RobFog** [comment]
One more thing: if you modify the commit message to say "fix #279391" and "fix 278153" that should close both issues in the tracker automatically when the pull request is merged. 😄

**Comment by pthvogt** [comment]
Had to correct the Volta Tempo test. Its inital version was wrong :-)

File: libmscore/layout.cpp [hunk 1]
```diff
@@ -3500,33 +3469,131 @@ System* Score::collectSystem(LayoutContext& lc)
...
-                        // can exist without chord or rest too.
-                        if (h->isLayoutInvalid())
-                              h->layout();
-                        toHarmony(e)->autoplaceSegmentElement(styleP(Sid::minHarmonyDistance));
+                  }
+            }
+
+      //-------------------------------------------------------------
+      // Harmony, 1st place
+      // If we have FretDiagrams, we want the Harmony above this and
```
[the remark this answers is not in the document]
Comment (anatoly-os) [reply libmscore/layout.cpp]: Current state is better, but far from ideal. screencast 2018-12-11 09-27-46

I suggest not to overcomplicate things and keep some default behaviour which could be incorrect in some cases, but is easy to implement. See my comment above.

**Comment by anatoly-os** [comment]
@pthvogt could you please separate the code to 2 commits. First one would fix the tempo playback issue. The second one would take care of the chords/frets positions.

**Comment by pthvogt** [comment]
The two commits should be still in one PR?

And: the behaviour of Fret + Harmony + Volta should be discussed in the issue, not in the PR :-)

**Comment by anatoly-os** [comment]
Yes, two commits in one PR. Actually, only the code fixing the particular issue should be in this PR. The improvement related to chords/frets should be indeed discussed and implemented in the scope of another issue and another PR.

File: libmscore/layout.cpp (hunk 1 above)
[the remark this answers is not in the document]
Comment (dmitrio95) [reply libmscore/layout.cpp]: It seems to be caused by not updating skylines after vertical aligning of voltas. This is not a problem of this patch but still needs resolving if we intend to place something above voltas. Try to update voltas skylines like it is done inside `processLines()`, and the autoplacement algorithm will handle them correctly.
https://github.com/musescore/MuseScore/blob/08b3ca9f6a694c312ab419c6164b958da3264b3f/libmscore/layout.cpp#L2971-L2972

File: libmscore/layout.cpp [hunk 2]
```diff
@@ -3500,33 +3469,131 @@ System* Score::collectSystem(LayoutContext& lc)
...
+                  for (Element* e : s->annotations()) {
+                        if (e->isHarmony()) {
+                              Harmony* h = toHarmony(e);
+                              // Layout of harmony seems to be bound currently
+                              // to ChordRest (see ChordRest::shape). But harmony
+                              // can exist without chord or rest too.
+                              if (h->isLayoutInvalid())
+                                    h->layout();
+                              toHarmony(e)->autoplaceSegmentElement(styleP(Sid::minHarmonyDistance));
+                              }
```
Comment (dmitrio95) [inline libmscore/layout.cpp]: Wound it perhaps be better to extract this to a separate function? It seems this chunk of code is just repeated twice, so any possible changes to them should be repeated which is easy to forget.

**Comment by pthvogt** [comment]
Changed PR to fix only #278153.

**Pull request merged**